src/token-service/metadata-token-service.ts

```diff
@@ -11,68 +11,102 @@ const DEFAULT_OPTIONS: Options = {
     private readonly url: string;
     private readonly opts: Options;
     private token?: string;
+    private tokenExpired = 0;
+    private tokenTimeToRefresh = 0;
+    private currentFetchToken?: Promise<string>;
+    private currentInitialize?: Promise<void>;
 
     constructor(url: string = DEFAULT_URL, options: Options = DEFAULT_OPTIONS) {
         this.url = url;
         this.opts = options;
     }
 
     async getToken(): Promise<string> {
-        if (!this.token) {
-            await this.initialize();
 
-            if (!this.token) {
-                throw new Error('Token is empty after MetadataTokenService.initialize');
+        if (!this.token || Date.now() >= this.tokenExpired) {
+            await this.initialize(); // cay throw error
+        } else if (Date.now() >= this.tokenTimeToRefresh) {
+            try {
+                this.token = await this.fetchToken();
+            } catch {
+                // nothing - use old token
             }
-
-            return this.token;
         }
 
-        return this.token;
+        return this.token as string;
     }
 
     private async fetchToken(): Promise<string> {
-        const res = await axios.get<{ access_token: string }>(this.url, this.opts);
+
+        // deduplicate real fetch token requests in any async case
+        if (!this.currentFetchToken) {
+            this.currentFetchToken = this._fetchToken().finally(() => {
+                delete this.currentFetchToken;
+            });
+        }
+
+        return this.currentFetchToken as Promise<string>;
+    }
+
+    private async _fetchToken(): Promise<string> {
+
+        const res = await axios.get<{ token_type: 'Bearer', access_token: string, expires_in: number, data: {} }>(this.url, this.opts);
 
         if (res.status !== 200) {
             throw new Error(`failed to fetch token from metadata service: ${res.status} ${res.statusText}`);
         }
 
+        const timeLeft = res.data.expires_in * 1000 - TOKEN_MINIMUM_LIFETIME_MARGIN_MS;
+
+        if (timeLeft <= 0) {
+            throw new Error('failed to fetch token: insufficient lifetime');
+        }
+
+        this.tokenExpired = Date.now() + timeLeft;
+        this.tokenTimeToRefresh = Date.now() + res.data.expires_in * 1000 * (1 - TOKEN_LIFETIME_LEFT_TO_REFRESH_PCT / 100);
+
         return res.data.access_token;
     }
 
     private async initialize() {
-        if (this.token) {
-            return;
+
+        // deduplicate initialize requests in any async case
+        if (!this.currentInitialize) {
+            this.currentInitialize = this._initialize().finally(() => {
+                delete this.currentInitialize;
+            });
         }
 
+        return this.currentInitialize as Promise<void>;
+    }
+
+    private async _initialize() {
         let lastError = null;
 
-        for (let i = 0; i < 5; i++) {
+        delete this.token;
+
+        for (let i = 0; i < MAX_ATTEMPTS_NUMBER_TO_GET_TOKEN_IN_INITIALIZE; i++) {
```

Comment:
Because if we already have a token and it has at least 15 min life left.  In case of an error, there is no need to retry, we can continue to use the old token.